### PR TITLE
feat(fast-element): add content composition

### PR DESCRIPTION
**Pull request by EisenbergEffect**
# Description

This feature adds the ability for expressions located in arbitrary HTML content to return a template to be rendered into that location.

## Motivation & context

For an overview of the feature, have a read through issue #3039. The technical approach outlined there was exactly what was put into place here. Additionally, some basic cache smarts were added to the content binding mechanism, enabling the entire `when` directive to be deleted and instead simply be *syntax sugar* on top of the low-level composition capabilities. Everything from #3039 is implemented in this PR with the exception of the changes to `repeat`. Those are planned for a separate PR next week.

Unit tests have been added around content composition, `repeat`, and other areas touched by this PR. As part of the unit test writing process, I realized it would be nice to have a helper to await the next dom batch update, so `DOM.nextUpdate(): Promise` was added.

The best place to look for an understanding of the new content composition features is the update to the documentation in this PR.

## Issue type checklist

- [ ] **Chore**: A change that does not impact distributed packages.
- [ ] **Bug fix**: A change that fixes an issue, link to the issue above.
- [x] **New feature**: A change that adds functionality.

Technically, this is a breaking change as it removes `WhenDirective` and `WhenBehavior`. However, to my knowledge, no one was using those directly. Instead, they were using the `when` helper in their templates. The `when` helper remains, but its implementation is now changed to leverage the core composition system.

## Process & policy checklist

- [x] I have added tests for my changes.
- [x] I have tested my changes.
- [x] I have updated the project documentation to reflect my changes.
- [x] I have read the [CONTRIBUTING](https://github.com/Microsoft/fast-dna/blob/master/CONTRIBUTING.md) documentation and followed the [standards](https://www.fast.design/docs/en/contributing/standards) for this project.

### CODE REVIEW

**Comment by awentzel** [comment]
Is `ACKNOWLEDGEMENT.md` a special file type?  Curious, why it's in all caps, yet the other files are not.

File: packages/web-components/fast-element/src/directives/when.ts [hunk 1]
```diff
@@ -1,77 +1,26 @@
...
+): CaptureType<T> {
+    let getTemplate: Expression<T, SyntheticViewTemplate>;
 
-    public createBehavior(target: any): WhenBehavior {
-        return new WhenBehavior(target, this.expression, this.template);
+    if (typeof templateOrTemplateExpression === "function") {
+        getTemplate = templateOrTemplateExpression;
+    } else {
+        getTemplate = (): SyntheticViewTemplate => templateOrTemplateExpression;
     }
```
Comment (janechu) [inline packages/web-components/fast-element/src/directives/when.ts:22]: nit: i'd suggest:
```typescript
const getTemplate: Expression<T, SyntheticViewTemplate> = typeof templateOrTemplateExpression === "function"
    ? templateOrTemplateExpression
    : (): SyntheticViewTemplate => templateOrTemplateExpression;
```

or:
```typescript
return (source: any, context: ExecutionContext): SyntheticViewTemplate | null =>
    !condition(source, context)
        ? null
        : typeof templateOrTemplateExpression === "function"
        ? templateOrTemplateExpression(source, context)
        : (): SyntheticViewTemplate => templateOrTemplateExpression(source, context);
```

Comment (EisenbergEffect) [reply packages/web-components/fast-element/src/directives/when.ts:22]: I'll refactor to the first one. The second would be a bit slower due to repeated type checks, where as the first option does it one time for all instances created from the template.

**Comment by EisenbergEffect** [comment]
@awentzel It's not special in the way that GitHub understands it, but it is a special file in the general sense that we clearly acknowledge other libraries, etc. that influence us. I picked up this practice 5ish years ago and have used it in my own open source. There's precedent in other projects as well. See Angular for another example: https://github.com/angular/angular/tree/master/docs

File: packages/web-components/fast-element/docs/guide/defining-elements.doc.md [hunk 2]
```diff
@@ -495,14 +511,12 @@ Some context properties are opt-in because they are more costly to update. So, f
 
 #### Composing Templates
 
-The `HTMLTemplate` returned from the `html` tag helper is also a directive itself. As a result, you can create templates and compose them into other templates.
+The `ViewTemplate` returned from the `html` tag helper has special handling when it is used inside of another template. The is that you can create templates and compose them into other templates.
```
Comment (nicholasrice) [inline packages/web-components/fast-element/docs/guide/defining-elements.doc.md:514]: > **The** is that you can create templates and compose them into other...


Should "the" be "this"? Phrasing seems off.

Comment (EisenbergEffect) [reply packages/web-components/fast-element/docs/guide/defining-elements.doc.md:514]: Definitely off 🤣  I'll fix it.

**Comment by nicholasrice** [comment]
Looks great!

**Pull request merged**